### PR TITLE
Update vector conversions

### PR DESCRIPTION
**Pull request by sriharshakandala**
Add custom `CovariantVector` to `Contravariant123Vector` transformation functions 
- [ ] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [ ] Unit tests are included OR N/A.
- [ ] Code is exercised in an integration test OR N/A.
- [ ] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by sriharshakandala** [comment]
bors try

**Review by charleskawczynski — COMMENTED** [review]
This PR looks good to me. It's a bit unfortunate that these are not captured in the axis tensor benchmarks. Can you point to the methods that these new functions are overriding / specializing on?

I'm curious if we can write a more general overloading, like we did before for `project / transform`. It seems that one crux of this is to cache `∂x∂ξ' * ∂x∂ξ`, which seems like a good idea (I forgot that we left that on the table before).

Related issue: #867

**Comment by sriharshakandala** [comment]
> This PR looks good to me. It's a bit unfortunate that these are not captured in the axis tensor benchmarks. Can you point to the methods that these new functions are overriding / specializing on?
> 
> I'm curious if we can write a more general overloading, like we did before for `project / transform`. It seems that one crux of this is to cache `∂x∂ξ' * ∂x∂ξ`, which seems like a good idea (I forgot that we left that on the table before).

In the context of the GPU version, we noticed that this function uses much less register space per thread compared to the pre-existing version. Looking at the Nsight Systems profile for the pre-existing version, the fundamental bottleneck seems to be register space. This helps us use more threads per block or enable the GPU to schedule multiple blocks on the same SM resulting in significant speedup.

**Comment by sriharshakandala** [comment]
> This PR looks good to me. It's a bit unfortunate that these are not captured in the axis tensor benchmarks. Can you point to the methods that these new functions are overriding / specializing on?
> 
> I'm curious if we can write a more general overloading, like we did before for `project / transform`. It seems that one crux of this is to cache `∂x∂ξ' * ∂x∂ξ`, which seems like a good idea (I forgot that we left that on the table before).

I believe we provide a specialization that overrides https://github.com/CliMA/ClimaCore.jl/blob/main/src/Geometry/conversions.jl#L102 for this specification. Please correct me if I am missing something!

**Comment by sriharshakandala** [comment]
bors try

**Comment by charleskawczynski** [comment]
https://github.com/CliMA/ClimaCore.jl/blob/main/src/Geometry/conversions.jl#L102 for this specific case. Please correct me if I am missing something!

Ah, yeah, that's a pretty bad method (it calls `project` twice). In fact, all of the generic ones look pretty bad (cc @bloops)

https://github.com/CliMA/ClimaCore.jl/blob/main/src/Geometry/conversions.jl#L73-L109

File: src/Geometry/conversions.jl [hunk 1]
```diff
@@ -25,6 +25,29 @@ LocalVector(
...
+
+ContravariantVector(
+    u::CovariantVector{T, I},
+    local_geometry::LocalGeometry{I},
+) where {T, I} = local_geometry.gⁱʲ * u
+
+CovariantVector(
+    u::ContravariantVector{T, I},
+    local_geometry::LocalGeometry{I},
+) where {T, I} = local_geometry.gᵢⱼ * u
```
Comment (charleskawczynski) [inline src/Geometry/conversions.jl:49]: Is diagonalization along `I` required here?

Comment (charleskawczynski) [reply src/Geometry/conversions.jl:49]: (just out of curiosity)

Comment (sriharshakandala) [reply src/Geometry/conversions.jl:49]: This works the full vector, with all `ndim` components. The first two deals with partial vectors!

Comment (charleskawczynski) [reply src/Geometry/conversions.jl:49]: Ok, I think you're saying yes. The generic method already defined does not diagonalize:

```julia
CovariantVector(
    u::ContravariantVector,
    local_geometry::LocalGeometry{I},
) where {I} = project(
    CovariantAxis{I}(),
    project(ContravariantAxis{I}(), u),
    local_geometry,
)
```

**Review by charleskawczynski — APPROVED** [review]
This is fine for now, but we should really improving the performance of these operators more systematically (like we did with `project` / `transform`).

File: src/Geometry/conversions.jl [hunk 2]
```diff
@@ -25,6 +25,29 @@ LocalVector(
...
+Contravariant123Vector(
+    u::CovariantVector{T, (1, 2)},
+    local_geometry::LocalGeometry{(1, 2, 3)},
+) where {T} = local_geometry.gⁱʲ * Covariant123Vector(u[1], u[2], zero(u[1]))
+
+Contravariant123Vector(
+    u::CovariantVector{T, (3,)},
+    local_geometry::LocalGeometry{(1, 2, 3)},
+) where {T} =
+    local_geometry.gⁱʲ * Covariant123Vector(zero(u[1]), zero(u[1]), u[1])
```
Comment (charleskawczynski) [inline src/Geometry/conversions.jl:38]: @sriharshakandala, can we hand-write this and avoid the dense mul?

**Pull request merged**